Coordinating wasm-bindgen versions

data-pup opened the pull request:
Fixes #146. This aims to solve the issue of making sure that there is (a) a locally installed `wasm-bindgen` and (b) that it is the same version as specified in the `Cargo.toml` file.

As of now, there are still some rough edges that I want to hone down here. It seemed like a good idea to check in the existing work on this issue that I got done today, to make sure that I am headed in the right direction on this. Any feedback is totally welcome!

I'll also close #81 now, since I believe this will also take care of #51.

I should be able to get this all wrapped up tomorrow, had to spend a little more time than expected today getting up to speed with the changes that have occurred since then. Loving all the stuff people have contributed! :smile: 

---

Make sure these boxes are checked! 📦✅

- [x] You have the latest version of `rustfmt` installed and have your 
      cloned directory set to nightly
```bash
$ rustup override set nightly
$ rustup component add rustfmt-preview --toolchain nightly
```
- [x] You ran `rustfmt` on the code base before submitting
- [x] You reference which issue is being closed in the PR text

✨✨ 😄 Thanks so much for contributing to wasm-pack! 😄 ✨✨

data-pup commented on src/bindgen.rs:
```diff
@@ -1,16 +1,40 @@
+        let stderr = String::from_utf8_lossy(&output.stderr).to_string();
+        let e = Error::Cli { message, stderr };
+        Err(e)
+    }
```
Not a giant fan of this :point_up: Creating the `Error:Cli` manually, rather than through the `cli(..)` function seems a little hacky, but the other approach caused a return type error, since this function returns a `Result<bool, Error>

mgattozzi replied:
Odd. What was the error?

data-pup replied:
If I use the `Error::cli` function here I end up with the following error:

```
expected type `std::result::Result<bool, _>`
found type `std::result::Result<(), _>`
```

This is happening because that function wraps the error up into a result, which is fine for functions that also return a `Result<(), Error>`, but means that it won't work for functions that return something else in the `Ok` case :sad`

```rust
pub fn cli(message: &str, stderr: Cow<str>) -> Result<(), Self> {
```

data-pup commented on src/bindgen.rs:
```diff
@@ -1,16 +1,40 @@
+
+pub fn wasm_bindgen_locally_installed(_path: &str) -> bool {
+    fs::metadata(LOCAL_BINDGEN_PATH).is_ok()
+}
```
Should I create an absolute path here? Or is the relative path in `LOCAL_BINDGEN_PATH` enough? I think I saw an issue regarding what happens when `wasm-pack` is run from within the `pkg` directory, so I was thinking that using the crate path could make this a little less fragile.

If relative path is satisfactory, I can remove the parameter here, otherwise an absolute path can be formed with `format!` :)

data-pup commented on src/command/init.rs:
```diff
@@ -188,9 +188,25 @@ impl Init {
+            info!(&log, "Installing wasm-bindgen-cli...");
+            bindgen::cargo_install_wasm_bindgen(&self.crate_path, &bindgen_version, step)?;
+            info!(&log, "Installing wasm-bindgen-cli was successful.");
+        }
```
The control flow here is the main thing I'd like to clean up, not especially happy with these nested `if` statements.

data-pup commented:
I think this is ready for review now :smile: Let me know if you have any feedback!

ashleygwilliams commented:
@data-pup awesome! i'll take a look soon- in the meantime it looks like rust fmt needs appeasing lolsob. thanks for putting so much effort into this! it's super appreciated. as mentioned previously- i think i'm down to sneak this in in a point release because it's gonna make everything way more awesome :)

data-pup commented:
Ahh pardon! I ran `fmt` before I rebased this, that's fixed now :) Happy I could help!

mgattozzi commented on src/bindgen.rs:
```diff
@@ -1,16 +1,46 @@
 use PBAR;
 
-pub fn cargo_install_wasm_bindgen(step: &Step) -> Result<(), Error> {
+static LOCAL_BINDGEN_PATH: &str = "bin/wasm-bindgen";
```
You should cfg this so that windows is "bin\\wasm-bindgen". While windows can handle it (sort of? windows pathing is wild technically the yen symbol is a valid path separator) it's best to have an option for it.

mgattozzi commented on src/bindgen.rs:
```diff
@@ -1,16 +1,46 @@
+static LOCAL_BINDGEN_PATH: &str = "bin/wasm-bindgen";
+
+fn wasm_bindgen_exists_locally(crate_path: &str) -> bool {
+    let path_str = format!("{}/{}", crate_path, LOCAL_BINDGEN_PATH);
```
Same reasoning here for the path.

mgattozzi commented on src/command/init.rs:
```diff
@@ -192,9 +192,25 @@ impl Init {
-        bindgen::cargo_install_wasm_bindgen(step)?;
-        info!(&log, "Installing wasm-bindgen-cli was successful.");
+        let bindgen_version =
+            manifest::get_wasm_bindgen_version(&self.crate_path)?.ok_or(Error::CrateConfig {
```
We should just make a `crate_config` function for `Error` and use that here.

mgattozzi commented on src/command/init.rs:
```diff
@@ -192,9 +192,25 @@ impl Init {
+            })?;
+
+        if bindgen_version.is_empty() {
+            let e = Error::CrateConfig {
```
And here!

mgattozzi replied:
then you can do `return Err(Error::crate_config("wasm-bindgen version dependency was empty"))`

mgattozzi commented on src/manifest.rs:
```diff
@@ -148,6 +148,13 @@ pub fn get_crate_name(path: &str) -> Result<String, Error> {
 }
 
+pub fn get_wasm_bindgen_version(path: &str) -> Result<Option<String>, Error> {
+    let version: Option<String> = read_cargo_toml(path)?
```
you shouldn't need to specify `Option<String>` here. The type is inferred as the return type

data-pup commented:
Added some more commits to fix the issues mentioned in @mgattozzi 's review. Thanks for the feedback! I added some platform specific path strings, and changed the `get_wasm_bindgen_version` function to return a `Result<String, Error>`. 

> then you can do return Err(Error::crate_config("wasm-bindgen version dependency was empty"))

Regarding this point, I realized that there was already a `crate_config` error, but did run into the same issue I'd seen with the other `cli` helper function. These were returning a `Result<(), Error>`, which meant that they aren't usable in a function that _does_ return a value in the success case.

I altered these so that they return `Self`, so that we can use these methods elsewhere too. This felt a little opinionated on my part, but I don't think it affects readability.

data-pup commented:
> You'll also need to make sure the part that calls wasm-bindgen in the build step uses the local one!

Still need to take care of this part, I'll do that right now though! :smiley:

data-pup commented:
Ready for another round of review, whenever you have time!

ashleygwilliams commented on src/bindgen.rs:
```diff
@@ -1,24 +1,67 @@
+    dep_version: &str,
+    step: &Step,
+) -> Result<bool, Error> {
+    let msg = format!("{}Checking WASM-bindgen dependency...", emoji::CHECK);
```
not sure if this should be a step displayed to users- i think we should just do it in the install wasm-bindgen step

ashleygwilliams commented on src/bindgen.rs:
```diff
@@ -1,24 +1,67 @@
 
-pub fn cargo_install_wasm_bindgen(step: &Step) -> Result<(), Error> {
+#[cfg(not(target_family = "windows"))]
+static LOCAL_BINDGEN_PATH: &str = "bin/wasm-bindgen";
```
this is local to the crate- correct? if so we'll want to add it to the .gitignore

ashleygwilliams commented on src/bindgen.rs:
```diff
@@ -1,24 +1,67 @@
+        let installed_version = s.trim();
+        Ok(installed_version == dep_version)
+    } else {
+        let error_msg = "Could not find version of local wasm-bindgen";
```
we may want to give some sort of instruction about how to handle this if it happens. feels a little like a dead end right now :)

data-pup commented:
I like your points regarding the `--no-installs` flag, that's an important detail I don't think my changes currently account for. Would it make sense in that case to search through the locations in the $PATH environment variable for wasm-bingen?

Overriding the local behavior with a flag makes sense to me, but right now I'm not sure if I have a solid answer to the story about how this works either. I will think about this too, but feel free to follow up if you come to a decision :)

ashleygwilliams commented:
@data-pup i also don't know exactly what we should do! very open to suggestions. i can take some time this weekend to think on it :)

mgattozzi commented:
Doesn't no installs imply using a local one anyways? From my understanding it won't be installing a global version. In fact wouldn't it make more sense to just use a sandboxed version anyways rather than relying on a global one? In tools like stack/cabal for Haskell it uses sandboxed versions of everything unless specified to use a global one which seems like the better choice, unless I'm misunderstanding what you mean by --no-installs @ashleygwilliams

data-pup commented:
Just wanted to follow up on this, do you happen to have any advice for next steps on my end? If there might still be some thought required regarding the interface etc., I'm happy to hold off on this!

I know I can definitely add some extra logic to handle the `--no-installs` flag, but a little unsure what that should generally look like at the moment.

Btw, thanks for being nice maintainers! Appreciate the feedback so far :)

data-pup commented:
I saw that #146 has been marked as Blocking Rust 2018, so I rebased this off of master to prepare for further work on this. At this point I think all of the changes requested by Michael have been taken care of, and I can take care of some of the minor details mentioned in Ashley's review today.

My current plan is to look for a globally installed version of `wasm-bindgen` that matches the dependency version if `--no-installs` is provided. If that cannot be found in the path, then it will fail rather than installing a new version.

If I should take a different route, let me know!

data-pup commented:
Took a swing at this today :) Let me know if you spot any problems!

mgattozzi commented:
@data-pup I'll have to take a look now that I'm back when I get the chance

data-pup commented:
Sounds good to me! I can also rebase this again, now that the path things from #220 have landed :+1:

fitzgen commented on src/bindgen.rs:
```diff
@@ -34,40 +61,112 @@ pub fn wasm_bindgen_build(
+    #[cfg(not(target_family = "windows"))]
+    let local_path = format!("{}/{}", crate_path, "bin/wasm-bindgen");
+    #[cfg(target_family = "windows")]
+    let local_path = format!("{}\\{}", crate_path, "bin\\wasm-bindgen");
```
If these functions were using `std::path::Path[Buf]`, then we wouldn't need these different code paths. Is there a reason we aren't using them?

data-pup replied:
No reason, I hadn't known about those. I can fix this :)

fitzgen commented on src/bindgen.rs:
```diff
@@ -3,25 +3,52 @@
+    // Determine if the `wasm-bindgen` dependency is already met for the given mode.
+    let dependency_met = if let Some(bindgen_path) = wasm_bindgen_path_str(path, install_permitted)?
+    {
+        wasm_bindgen_version_check(&bindgen_path, version)?
```
I *think* we don't want to propagate errors here, but instead treat them as "false, dependency not met".

fitzgen commented on src/bindgen.rs:
```diff
@@ -34,40 +61,112 @@ pub fn wasm_bindgen_build(
+    crate_path: &str,
+    install_permitted: bool,
+) -> Result<Option<String>, Error> {
+    if install_permitted {
```
If we are not permitting installs, we should still allow *using* a local install, shouldn't we? Just not installing the local version? Or is it more of "force use of the global version and don't install it if missing"?

fitzgen commented on src/bindgen.rs:
```diff
@@ -34,40 +61,112 @@ pub fn wasm_bindgen_build(
+    let path = env::var("PATH")?;
+    for path_dir in path.split(path_sep) {
+        let prog_str = format!("{}/wasm-bindgen", path_dir);
+        if path::Path::new(&prog_str).is_file() {
```
I think we should just use https://crates.io/crates/which instead of rolling our own implementation.

fitzgen commented on src/bindgen.rs:
```diff
@@ -34,40 +61,112 @@ pub fn wasm_bindgen_build(
+    if output.status.success() {
+        let s = String::from_utf8_lossy(&output.stdout);
+        let installed_version = s.trim();
+        Ok(installed_version == dep_version)
```
Running `wasm-bindgen --version` doesn't produce *just* the version number:

```
$ wasm-bindgen --version
wasm-bindgen 0.2.15
```

So I think we need to do a little bit more parsing here than just `.trim()`, since we are getting the expected version from `Cargo.toml` which *will* just be the version number.

data-pup replied:
Ah! Good catch. I can take care of that.

fitzgen commented on src/error.rs:
```diff
@@ -44,23 +49,24 @@ pub enum Error {
     /// Construct a CLI error.
-    pub fn cli(message: &str, stderr: Cow<str>) -> Result<(), Self> {
-        Err(Error::Cli {
+    pub fn cli(message: &str, stderr: Cow<str>) -> Self {
```
These changes to the `Error` constructors are making this PR a bit noisier than it otherwise would be; unless this was requested for this PR, I'd have personally left them for a separate PR.

data-pup replied:
I agree that these are a little noisy, I can totally remove these changes. The tldr of why they are there is that these methods were causing type errors when I tried to use them from within a function that didn't also return a `Result<(), Error>`, but you are right that making that change separately is probably wiser :)

data-pup commented:
Thanks for the feedback @fitzgen! I can resolve the merge conflicts today, and take care of some of your inline comments as well. I can have something EOD, since this relates to a 'Blocking Rust 2018' issue :crab:

data-pup commented:
Continuing this work in #244, because the merge conflicts got a little too drastic.
